### ISSUE TITLE
om_minverser.exe crash

### ISSUE THREAD

**Issue by MartinFRE** [openmeeg/openmeeg#76]
Hi,

I am trying to compute a head model with the brainstorm toolbox after segmenting the MRI with BrainSuite, following this tutorial: http://neuroimage.usc.edu/brainstorm/Tutorials/TutBem; thereby using the openmeeeg software.

No matter if I enable "use adjoint formulation" or "process dipoles by blocks of 4000" or reducing the number of vertices, I always get this error after ~ 1 min:
- Windows error message "om_minverser.exe has stopped working"
- followed by a brainstorm error message (screenshot attached here)

I don't think it's a memory issue (I am working on a server with 256 GB RAM).
- System: Windows 8 Pro, Intel Xeon CPU E5-2690 0 @ 2.90GHz, 64 bit, 256 GB RAM

I tried to use different versions of OpenMEEG, but still get the same error.

Right now I don't know how to solve this problem further. Do you have any ideas? Thanks!

Best,
Martin

**Comment by agramfort** [comment]
can you share the BEM files? How many layers are you using?

**Comment by papadop** [comment]
In addition, are you using the latest version of BrainStorm. There has been some changes recently for issues with the files generated by om_assemble when those are big > 4Gb. In essence, the files should be matlab files (.mat) instead of .bin files (recent versions of brainstorm have been changed to do that).

Just to be sure, verify that you are using a 64 bits version of brainstorm/openmeeg (otherwise openmeeg wont be able to use the 256 Gb).

**Comment by MartinFRE** [comment]
Thanks for the fast answers! 

I am using following versions:
- OpenMEEG-2.2.0-win64-x86_64-cl-OpenMP-shared 
- Brainstorm Version 3.2 (11-Sep-2015)
- Matlab: Tested in 2014a & 2014b, both 64bit

Here are the BEM files I'm using (it's also the same error for other subjects):
https://www.dropbox.com/s/n346w7tchdgntp4/BEM.zip?dl=1

I tried it with different numbers of layers (default was 17 for volume model, but even with only 2 layers I get the error).

By the way I forgot to attach the screenshot of the error, here it is:
![error minverser](https://cloud.githubusercontent.com/assets/14314969/9909341/15ce246c-5c98-11e5-8d7f-68e374b1bce2.png)

**Comment by laurajanem** [comment]
Hi,
I am having the same problem when trying to use OpenMEEG through Brainstorm.

I am using the latest version of Brainstorm: Version 3.2, Feb. 24, 2016
Matlab: 2015b
Windows 7 Professional, Intel Xeon CPU E5-2630 vs @2.40 GHz (2 processors) 64.0 GB RAM
Latest version of OpenMEEG: OpenMEEG-2.2.0-win64-x86_64-cl-OpenMP-shared

I don't think this is due to the .bin/.mat problems previously described.  All of the files being generated are .mat, and stepping through the Brainstorm Matlab code, I see that it is running the version of the bst_openmeeg.m file that explicitly saves everything in .mat file format.  

I also don't think it is a file size issue, since I've been purposefully testing it with small numbers of vertices in the BEM meshes.

Depending on which option I choose (adjoint or adaptive), the program crashes at different parts of the process.

For the adjoint case, it crashes on om_gain and this is the corresponding log:

om_assemble version 2.2.0 (exported) compiled at Feb 19 2013 14:01:10


| ------ om_assemble
| -H2EM
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg.geom
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg.cond
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg_loc_eeg.txt
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg_h2em.mat
| -----------------------
Sorted List : 	1 0 2 
Sorted Domains : 	Brain	Skull	Scalp	Air	
Total number of points    : 1926
Total number of triangles : 3840

Checking
Mesh 0 : internal conductivity = 1 and external conductivity = 0.0125
Mesh 1 : internal conductivity = 0.0125 and external conductivity = 1
Mesh 2 : internal conductivity = 1 and external conductivity = 0

 Elapsed Time: 0.091 s.

om_gain version 2.2.0 (exported) compiled at Feb 19 2013 14:00:34


| ------ om_gain
| -EEGadjoint
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg.geom
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg.cond
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg_dipoles.txt
| C:\Users\ljm2331\Documents\Brainstorm_Analysis\brainstorm_db\Protocol01\anat\@default_subject\tess_innerskull_bem_642V_openmeeg.mat
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg_h2em.mat
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg_gain_eeg.mat
| -----------------------
Sorted List : 	1 0 2 
Sorted Domains : 	Brain	Skull	Scalp	Air	
Total number of points    : 1926
Total number of triangles : 3840

Checking
	Mesh 0 : internal conductivity = 1 and external conductivity = 0.0125
	Mesh 1 : internal conductivity = 0.0125 and external conductivity = 1
	Mesh 2 : internal conductivity = 1 and external conductivity = 0

**Exception: This matlab file does not contains a 2D full double Matrix as expected. Doing my best....**

----------------------
For the adaptive version, it crashes during om_minverser, with the following log file:

om_assemble version 2.2.0 (exported) compiled at Feb 19 2013 14:01:10


| ------ om_assemble
| -HM
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg.geom
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg.cond
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg_hm.mat
| -----------------------
Sorted List : 	1 0 2 
Sorted Domains : 	Brain	Skull	Scalp	Air	
Total number of points    : 1926
Total number of triangles : 3840

Checking
	Mesh 0 : internal conductivity = 1 and external conductivity = 0.0125
	Mesh 1 : internal conductivity = 0.0125 and external conductivity = 1
	Mesh 2 : internal conductivity = 1 and external conductivity = 0
OPERATOR S... (arg : mesh m1, mesh m2)
OPERATOR S... (arg : mesh m1, mesh m2)
OPERATOR S... (arg : mesh m1, mesh m2)
OPERATOR N... (arg : mesh m1, mesh m2)
OPERATOR N... (arg : mesh m1, mesh m2)
OPERATOR N... (arg : mesh m1, mesh m2)
OPERATOR D... (arg : mesh m1, mesh m2)
OPERATOR D... (arg : mesh m1, mesh m2)
OPERATOR D... (arg : mesh m1, mesh m2)
OPERATOR D... (arg : mesh m1, mesh m2)
OPERATOR S... (arg : mesh m1, mesh m2)
OPERATOR S... (arg : mesh m1, mesh m2)
OPERATOR N... (arg : mesh m1, mesh m2)
OPERATOR N... (arg : mesh m1, mesh m2)
OPERATOR D... (arg : mesh m1, mesh m2)

| Elapsed Time: 16.124 s.


om_minverser version 2.2.0 (exported) compiled at Feb 19 2013 13:59:57


| ------ om_minverser
| C:\Users\ljm2331\.brainstorm\tmp\openmeeg_hm.mat
| C:\Users\ljm2331\Documents\Brainstorm_Analysis\brainstorm_db\Protocol01\anat\@default_subject\tess_innerskull_bem_642V_openmeeg.mat
| -----------------------

----------------------------------------
So, it is getting through om_assemble without crashing, but I believe the problem is really with the format of the data in the .hm file that is generated.  The openmeeg_hm.mat file contains one variable called symmatrix, which is a structure with two fields: symmatrix.size and symmatrix.data

For the sample dataset I am using, symmatrix.size = 4486, and symmatrix.data is a 10064341 x 1 vector of doubles.  (Total file size is about 64 Mb).

Is om_gain, and somewhere internal to om_minverser expecting the data in the .hm file to be of a different format?  The error message generated by the crash during om_gain suggests this:
**Exception: This matlab file does not contains a 2D full double Matrix as expected. Doing my best....**

Also, when I try to run om_matrix_info from the command line on the openmeeg_hm.mat file, I get the same error about it not being a 2D full matrix, and then om_matrix_info crashes.

Here is the sample openmeeg_hm.mat file:
https://www.dropbox.com/s/o9sln6f8iegsuuq/openmeeg_hm.mat?dl=0

Thanks so much for your help!

**Comment by agramfort** [comment]
can you also share these files:

| C:\Users\ljm2331.brainstorm\tmp\openmeeg.geom
| C:\Users\ljm2331.brainstorm\tmp\openmeeg.cond

as well as the .tri files in the same folder?

thanks

**Comment by laurajanem** [comment]
Added the files here:

https://www.dropbox.com/sh/epidf63vx1rwops/AACOasxGX7NNpzF8nXA6mSCTa?dl=0

These were all generated through the Brainstorm Matlab interface.

Thanks!

**Comment by agramfort** [comment]
thanks. I managed to reproduce the pb. I'll try to have a look ASAP

or maybe @papadop can have a look too.

thanks for the bug report

**Comment by papadop** [comment]
Thank's for the report. Alexandre how did you reproduce the problem. I cannot with the trunk.

sipadan-> ../OpenMEEG/build/tools/om_matrix_info -sym -i openmeeg_hm.mat
../OpenMEEG/build/tools/om_matrix_info version 2.3.dev (Unversioned directory) compiled at Jan 27 2016 15:08:39

Loading : openmeeg_hm.mat
Dimensions : 4486 x 4486
Min Value : -0.0103983 (4011,4011)
Max Value : 0.00235591 (2128,4050)
First Values
-0.00456105 0.000276311 0.000279566 0.000113986 0.000246293 
-0.00511295 5.64053e-05 0.000302956 0.000307797 
-0.00518512 0.000195422 0.000312761 
-0.00541568 6.30971e-05 
-0.00529717 

../OpenMEEG/build/src/om_minverser openmeeg_hm.mat toto.mat
../OpenMEEG/build/src/om_minverser version 2.3.dev (Unversioned directory) compiled at Jan 27 2016 15:08:35


| ------ ../OpenMEEG/build/src/om_minverser
| openmeeg_hm.mat
| toto.mat
| -----------------------
-------------------------------------------
| Elapsed Time: 33.7263 s.
-------------------------------------------

So, if I did not screw the test, it seems to be already corrected on the trunk....
The correction probably came with an upgrade of matio (and there are other in the pipe) or I vaguely remember fixing a bug with symmetric matrices a while ago. 

Alexandre, with which version (and command) did you reproduce the bug ??

**Comment by laurajanem** [comment]
If the problem seems to have been fixed in the current version of the source code, could you possibly provide an updated Windows binary I could install on my machine? The most recent one available for download on the OpenMEEG site is from Feb. 19, 2013.

**Comment by dang-kai** [comment]
Here is the windows binary I recently compiled, see if it works for you: https://www.dropbox.com/s/ctbzzkpsfahytz7/Release_2016-03-05.zip?dl=0
And install the runtime libraries if necessary: https://www.microsoft.com/en-us/download/details.aspx?id=48145

**Comment by agramfort** [comment]
my bad. I cannot replicate the pb if I don't forget the the -sym param in
om_matrix_info

if these binaries work we should put them in the openmeeg download folder
on the gforge

**Comment by FabienneCologne** [comment]
Have you been able to solve this error? If yes, I would appreciate any advice... ;)

**Comment by agramfort** [comment]
The only option I see is try to recompile for your system:

https://github.com/openmeeg/openmeeg-project

@papadop any recommendation for windows?

**Comment by laurajanem** [comment]
Sorry for taking so long to update this.

I tried for a very long time to get the current version of OpenMEEG to compile on my Windows machine, and just couldn't get it to work. There are many reasons for this which I think are better suited to discussion in a different thread.

However, the Windows binaries provided by @dang-kai actually worked for me. Thank you so much for doing this!

For individuals having trouble running OpenMEEG on Windows through Brainstorm - replace the OpenMEEG binaries Brainstorm installs in the default location (something like C:\Users\xxx\.brainstorm\openmeeg\win64) with the ones provided by @dang-kai above, and it should work. 

Slightly off topic, but I encountered the same kinds of errors when trying to run OpenMEEG through the Matlab FieldTrip toolbox as well. Haven't tested the new binaries in that context yet, though.

Request - would it be possible to provide a set of updated precompiled Windows binaries using the current version of the toolbox?  

Thanks so much for your help!

**Comment by MartinFRE** [comment]
Thanks, @dang-kai's binaries worked for me, too!

Updated precompiled Windows binaries would indeed be nice!

**Comment by KLehongre** [comment]
Hello,

I have the same issue. Are by any chance the precompiled Windows binaries still available somewhere? The link from  @dang-kai doesn't work any more. 
I've tried to compile on my windows, but it failed...

Thank you for your help,
katia

**Comment by dang-kai** [comment]
Here is the fixed link: https://www.dropbox.com/s/4cv61zatcijgc0c/OpenMEEG_Release_2016-03-05.zip?dl=0

**Comment by KLehongre** [comment]
Thanks a lot!
Unfortunately, it still doesn't work for me, but the error is different. Now it's: 


***************************************************************************
** Error: OpenMEEG call: om_assemble -HM
**       "C:\Users\Katia LEHONGRE\.brainstorm\tmp\openmeeg.geom"
**       "C:\Users\Katia LEHONGRE\.brainstorm\tmp\openmeeg.cond"
**       "C:\Users\Katia LEHONGRE\.brainstorm\tmp\openmeeg_hm.mat"
** OpenMEEG error #1: 
** om_assemble version 2.3.dev (unknown) compiled at Mar  5 2016 11:27:34
** 
** | ------ om_assemble
** | -HM
** | C:\Users\Katia LEHONGRE\.brainstorm\tmp\openmeeg.geom
** | C:\Users\Katia LEHONGRE\.brainstorm\tmp\openmeeg.cond
** | C:\Users\Katia LEHONGRE\.brainstorm\tmp\openmeeg_hm.mat
** | -----------------------
** (DEPRECATED) Please consider updating your geometry file to the new format 1.1 (see data/README.rst): C:\Users\Katia LEHONGRE\.brainstorm\tmp\openmeeg.geom
** IO: load: Unknown mesh file format for C:\Users\Katia
** 
** Many OpenMEEG crashes are due to a lack of memory.
** Reduce the number of vertices on each BEM layer and try again.
***************************************************************************

I'm testing with the TutorialEpilepsy dataset from brainstorm. Do you think it comes from the dataset or from the binaries?

Thank you for your help,
best
katia

**Comment by dang-kai** [comment]
Maybe it's because you have a space in your input file path ( between "Katia" and  "LEHONGRE") that made openmeeg misunderstood.
Could you move files to another path without space, try again and post the result?

**Comment by KLehongre** [comment]
Thanks, yes it works without gap!

But it would be great if it could work with a gap in the path name because brainstorm's script use the user's home directory to store those files. 
To make it work, I have changed the bst_get function from brainstorm line 304 to specify where to create the .brainstorm folder. More precisely I've changed "bstUserDir = bst_fullfile(bst_get('UserDir'), '.brainstorm');" to "bstUserDir = bst_fullfile('D:', '.brainstorm');"

But I'm not sure that it won't create any problem for other functions...

Best,
katia
